### ISSUE TITLE
Save/restore for TensorGraph

### ISSUE THREAD

**Issue by peastman** [deepchem/deepchem#768]
In the past, our models have generally had a persistent Session object containing all the variables for the model.  They automatically wrote out checkpoints during training, and provided a `restore()` method to load the variables from the most recent checkpoint into memory.

TensorGraph works differently.  It creates a new Session every time you call `fit()` or `predict()`, loads the variables from the latest checkpoint, then throws away the Session before returning.  I find this behavior has a couple of problems.

First, it's really slow.  If you want to run prediction thousands of times, that means it will load the variables from disk thousands of times.  This is why for both RL and MAML I've basically had to subvert the design of TensorGraph.  I use it to build the Tensorflow graph, but when I need to do any calculations I ignore the corresponding parts of TensorGraph, pull out the needed internal fields, and do the calculations directly.

It also is inflexible, since it enforces that prediction can only ever be done based on the variables in the latest checkpoint file, no others.  A good example of where this causes problems is MAML.  It uses TensorGraph to define the model, but the actual optimization is done with a different optimizer and a different loss function.  That produces a model (which gets saved to disk) that is designed to be easy to train, but isn't optimized for any particular task.  To use it for prediction, you first do a few steps of gradient descent to produce a fine tuned version of the model (which does *not* get saved to disk), and do the prediction based on that.  But I can't use any of TensorGraph's prediction methods for it, because it would just throw away the tuned variables and replace them with the generic ones from disk.

I think we should consider changing this behavior.

**Comment by rbharath** [comment]
I'd be in favor of changing the behavior. Do you have thoughts on what the new save/restore API should behave?

**Comment by peastman** [comment]
I'd suggest something roughly identical to what's currently in A3C.  When you first create the object, it creates a persistent Session.  There's a `restore()` method to load the most recent checkpoint from disk.  I also included a `restore` argument to `fit()` that you can use to have it automatically load the latest checkpoint and continue training from where it left off.

**Comment by rbharath** [comment]
This sounds solid to me.

@lilleswing Do you see any potential issues with the proposed design?

**Comment by lilleswing** [comment]
SGTM should provide significant speedup.  We should make a statement about explicitly closing the session once one is done with a TensorGraph if we have a long running session.

**Comment by rbharath** [comment]
SGTM to me as well.

It would also be useful to have an option that explicitly turns off checkpointing. This is useful when hyperparameter searching.

**Comment by peastman** [comment]
Ok, sounds good.  I can implement this, but I'm going to wait until #763 is done, since they'll involve changes to a lot of the same code.

**Comment by patrickhop** [comment]
I encountered this when testing the approximate error bars. The speedup will be very welcome in that application.

**Comment by vid33** [comment]
Is there an elegant way to load from a particular checkpoint, rather than just the latest? Before the above changes were implemented, setting (if I remember correctly) model.global_step or model.latest_checkpoint would cause model.evaluate to actually evaluate from that particular checkpoint. This is no longer the case. It's of course possible to just load using tf functions. However, this is a little more cumbersome. Plus, while not a big deal, is it a good thing that it's possible to define model.latest_checkpoint, but have model.evaluate actually use other weights (I guess .build() is not run before .evaluate()).

**Comment by rbharath** [comment]
Re-opening this issue since question will not be visible otherwise.

**Comment by peastman** [comment]
How about this?

1. Add an optional argument to `restore()` that lets you specify the checkpoint file to load.
2. Add a `get_checkpoints()` method that returns a list of all checkpoint files for the model, ordered from most recent to oldest.

**Comment by vid33** [comment]
FWIW, the first thing I did was check whether 1. is possible.
